### ISSUE TITLE
Upgraded to 6.0 Logstash - Spammed with Could not retrieve remote IP address for beats input.

### ISSUE THREAD

**Issue by sc-chad** [logstash-plugins/logstash-input-beats#269]
In AWS using Amazon Linux for most instances.

My filebeat agents aren't upgraded yet they are still on 5.6.4.

Uninstalled previous logstash agent of 5.2.1 and did a clean install of 6.0.

Logstash input
`  beats {
    port => 5044
  }`

Results in hundreds of 
`[2017-11-17T00:52:42,143][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input.`

**Comment by wf744** [comment]
I have the same issue, how to solve it?

**Comment by sc-chad** [comment]
I reverted back to 5.2.0 again.  These last few versions feel like they completely skipped QA.

**Comment by sc-chad** [comment]
If you want to stay on 6.0 you can do this:

Going to:  

```
/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-input-beats-5.0.2-java/lib/logstash/inputs/beats/message_listener.rb
```

And deleting: 
```
begin
    hash.get("@metadata").put("ip_address", ctx.channel().remoteAddress().getAddress().getHostAddress())
  rescue #should never happen, but don't allow an error here to stop beats input
    input.logger.warn("Could not retrieve remote IP address for beats input.")
end
```

Fixes the issue.

**Comment by jordansissel** [comment]
Can you provide more details? I tried to reproduce this locally and was unsuccessful (I see no warnings).

My test setup with logstash and packetbeat:

* `bin/logstash -e 'input { beats { port => 5000 } }' (Logstash 6.0.0)
* `sudo ./packetbeat` (Packetbeat 6.0.0)

Packetbeat data is successfully transmitted and I see no warnings.

**Comment by sc-chad** [comment]
Absolutely @jordansissel will help anyway I can.

We are using Filebeat from the official repos on Amazon Linux. (Same with Logstash)

An example of one of our prospectors.

```
---
filebeat:
  prospectors:
    - input_type: log
      paths:
        - /var/log/secure
      encoding: plain
      fields:
        environment: production
        product: template
        role: base
        log_type: syslog
      fields_under_root: true
      document_type: syslog
      scan_frequency: 10s
      harvester_buffer_size: 16384
      max_bytes: 10485760


      tail_files: false

      # Experimental: If symlinks is enabled, symlinks are opened and harvested. The harvester is openening the
      # original for harvesting but will report the symlink name as source.
      #symlinks: false

      backoff: 1s
      max_backoff: 10s
      backoff_factor: 2

      # Experimental: Max number of harvesters that are started in parallel.
      # Default is 0 which means unlimited
      #harvester_limit: 0

      ### Harvester closing options

      # Close inactive closes the file handler after the predefined period.
      # The period starts when the last line of the file was, not the file ModTime.
      # Time strings like 2h (2 hours), 5m (5 minutes) can be used.
      close_inactive: 5m


      # Close renamed closes a file handler when the file is renamed or rotated.
      # Note: Potential data loss. Make sure to read and understand the docs for this option.
      close_renamed: false

      # When enabling this option, a file handler is closed immediately in case a file can't be found
      # any more. In case the file shows up again later, harvesting will continue at the last known position
      # after scan_frequency.
      close_removed: true

      # Closes the file handler as soon as the harvesters reaches the end of the file.
      # By default this option is disabled.
      # Note: Potential data loss. Make sure to read and understand the docs for this option.
      close_eof: false

      ### State options

      # Files for the modification data is older then clean_inactive the state from the registry is removed
      # By default this is disabled.
      clean_inactive: 0

      # Removes the state for file which cannot be found on disk anymore immediately
      clean_removed: true

      # Close timeout closes the harvester after the predefined time.
      # This is independent if the harvester did finish reading the file or not.
      # By default this option is disabled.
      # Note: Potential data loss. Make sure to read and understand the docs for this option.
      close_timeout: 0
```

Can't say for sure this specific prospector is the one causing it, but all of them are pretty close in configuration.

Logstash configuration was just the default.  Log input from filebeat port 5044 or whatever.  No filters when I was testing it.

**Comment by ph** [comment]
@sc-chad This is strange, would you mind doing the following and adding the new logging message to this issue?

```ruby
begin
    hash.get("@metadata").put("ip_address", ctx.channel().remoteAddress().getAddress().getHostAddress())
  rescue => e #should never happen, but don't allow an error here to stop beats input
    input.logger.warn("Could not retrieve remote IP address for beats input.", :message => e)
end

**Comment by wf744** [comment]
@sc-chad  after deleted:
```ruby
begin
    hash.get("@metadata").put("ip_address", tx.channel().remoteAddress().getAddress().getHostAddress())
  rescue #should never happen, but don't allow an error here to stop beats input
    input.logger.warn("Could not retrieve remote IP address for beats input.")
end
```

and run for a while, i got an other issue:
```
[2017-11-21T01:53:30,451][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T01:58:30,529][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:03:30,610][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:08:30,676][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:13:30,737][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:18:30,783][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:23:30,837][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:28:30,890][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:33:30,939][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:38:37,987][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:43:39,041][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:48:46,101][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:53:56,151][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T02:58:56,200][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T03:03:59,258][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T03:09:01,315][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T03:14:05,371][INFO ][logstash.filters.translate] refreshing dictionary file
[2017-11-21T03:19:05,426][INFO ][logstash.filters.translate] refreshing dictionary file
```

after that, Logstash do not work any way

**Comment by wf744** [comment]
@ph when i change the code as you show, ` :message => e` ,logstash emit the errors as follow:
```
[2017-11-21T16:19:51,914][WARN ][io.netty.channel.AbstractChannelHandlerContext] An exception 'java.lang.NullPointerException' [enable DEBUG level for full stacktra
ce] was thrown by a user handler's exceptionCaught() method while handling the following exception:
org.jruby.exceptions.RaiseException: (NameError) undefined local variable or method `e' for #<LogStash::Inputs::Beats::MessageListener:0x364993cc>
        at usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.logstash_minus_input_minus_beats_minus_5_dot_0_dot_2_minus_java.lib.logstash.inputs.beats.mess
age_listener.onNewMessage(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-input-beats-5.0.2-java/lib/logstash/inputs/beats/message_listener.rb:35) ~[?:?
]
[2017-11-21T16:19:54,174][WARN ][io.netty.channel.AbstractChannelHandlerContext] An exception 'java.lang.NullPointerException' [enable DEBUG level for full stacktra
ce] was thrown by a user handler's exceptionCaught() method while handling the following exception:
org.jruby.exceptions.RaiseException: (NameError) undefined local variable or method `e' for #<LogStash::Inputs::Beats::MessageListener:0x364993cc>
        at usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.logstash_minus_input_minus_beats_minus_5_dot_0_dot_2_minus_java.lib.logstash.inputs.beats.mess
age_listener.onNewMessage(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-input-beats-5.0.2-java/lib/logstash/inputs/beats/message_listener.rb:35) ~[?:?
]
[2017-11-21T16:19:54,935][WARN ][io.netty.channel.AbstractChannelHandlerContext] An exception 'java.lang.NullPointerException' [enable DEBUG level for full stacktra
ce] was thrown by a user handler's exceptionCaught() method while handling the following exception:
org.jruby.exceptions.RaiseException: (NameError) undefined local variable or method `e' for #<LogStash::Inputs::Beats::MessageListener:0x364993cc>
        at usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.logstash_minus_input_minus_beats_minus_5_dot_0_dot_2_minus_java.lib.logstash.inputs.beats.mess
age_listener.onNewMessage(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-input-beats-5.0.2-java/lib/logstash/inputs/beats/message_listener.rb:35) ~[?:?
]

```

**Comment by robbavey** [comment]
@wf744 I suspect part of code change @ph suggested was missed:

Where the previous code stated

```
rescue #should never happen, but don't allow an error here to stop beats input
```

This should be changed to 

```
rescue => e
```

as well as the change to the existing `input.logger.warn' statement

Would you mind updating to include this line, and post the results?

Additionally, can you give us some more information about your setup? Is it also in AWS, using Amazon Linux? If so, is it part of a VPC, a regular EC2 instance, or something else?

Thanks

**Comment by ph** [comment]
@robbavey I am looking at @sc-chad new log statement, I was not expecting anything from this line to be NULL, I didn't run into this locally so its surely linked to the environment.

```
 tx.channel().remoteAddress().getAddress().getHostAddress())
```

**Comment by sc-chad** [comment]
@ph Sorry I haven't posted any new logs with 6.0.  I'll do that today.

**Comment by wf744** [comment]
@robbavey , sorry, i missed  part of code, i will add the code and test again.
It seems that this issue only appear in peak messaging times with my test.
My environment is 
```
LSB Version:    :base-4.0-amd64:base-4.0-noarch:core-4.0-amd64:core-4.0-noarch
Distributor ID: CentOS
Description:    CentOS release 6.5 (Final)
Release:        6.5
Codename:       Final
```

```
Linux iZwz9514uorf593pps5nioZ 2.6.32-431.23.3.el6.x86_64 #1 SMP Thu Jul 31 17:20:51 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

```
filebeat6.0
logstash6.0
Elasticsearch6.0
kibana6.0
x-pack6.0
```
I use aliyun.com 's ECS

**Comment by sc-chad** [comment]
Just to be clear you want me to replace with:

```
begin
    hash.get("@metadata").put("ip_address", ctx.channel().remoteAddress().getAddress().getHostAddress())
  rescue => e #should never happen, but don't allow an error here to stop beats input
    input.logger.warn("Could not retrieve remote IP address for beats input.", :message => e)
end
```

**Comment by sc-chad** [comment]
Sorry holidays are getting busy.  I've added the above code to two of our parsers.  I'll post back with logs as soon as I can.

**Comment by sc-chad** [comment]
Well that didn't take long.  Here ya go.

```
[2017-11-22T17:01:11,159][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input. {:message=>#<NoMethodError: undefined method `getAddress' for nil:NilClass>}
[2017-11-22T17:01:11,162][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input. {:message=>#<NoMethodError: undefined method `getAddress' for nil:NilClass>}
```

**Comment by robbavey** [comment]
@sc-chad I've put up a PR for the original issue. I think the second problem you describe is unrelated to the original problem, would you mind creating a separate issue for that, including all the details?

Thanks

**Comment by sc-chad** [comment]
@robbavey if you want me to add more details to help with that PR I'll be more than happy to.

I'll create another issue for the stack trace sometime this week.  Thanks again!

**Comment by robbavey** [comment]
@sc-chad That would be great!

Any more information about your setup would be helpful too (leaving out any sensitive information!) in helping us reproduce and fix your issue:
 - Do you have a sense as to whether this is happening for every event, most events, or a small percentage of events? (If it helps, events that do not suffer from this error will have the `ip_address` field populated in the event metadata field, and those that do will not)
 -  A short description of your setup - the more info the better for us to troubleshoot. 

Thanks

**Comment by BrianMBeaulieu** [comment]
I'm also having this issue.
tar distribution of 6.0.0.

input {
  beats {
    port => 5044
    type => "beats"
  }
}

**Comment by BrianMBeaulieu** [comment]
Almost all of my beat agents are < 6.0.0. I am not sure if that matters.
Logstash is 6.0.0.

**Comment by robbavey** [comment]
@BrianMBeaulieu Couple of follow up questions:

- What is the platform(s) you are running logstash and beats on?
- Are you running in the cloud?
- Do you have a sense as to what percentage of events this is happening on?

**Comment by sc-chad** [comment]
@robbavey no problem.  I'm out of office all week at re:invent, but i'll check this first thing next week.

**Comment by BrianMBeaulieu** [comment]
@robbavey 
Logstash - Amazon Linux
Linux es-broker 4.9.51-10.52.amzn1.x86_64 #1 SMP Fri Sep 29 01:16:19 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux


Beats - Mostly Windows (winlogbeat, filebeat).

It appears to be an issue only with filebeat for json data i'm pulling in from some of our application logs..  as winlogbeat indexes have all their data but the filebeat indexes don't..

**Comment by BrianMBeaulieu** [comment]
I'm not sure it matters, but this server is behind an iptables firewall, being NAT'd..

**Comment by robbavey** [comment]
@BrianMBeaulieu Thanks for all the info - these are all useful data points.

**Comment by BrianMBeaulieu** [comment]
[2017-11-29T13:16:20,854][INFO ][org.logstash.beats.Server] Starting server on port: 5044
[2017-11-29T13:18:25,953][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5044, remote: 10.100.1.99:54498] Exception: Connection reset by peer
[2017-11-29T13:19:00,369][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5044, remote: 10.100.1.99:54505] Exception: Connection reset by peer
[2017-11-29T13:26:03,307][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input.

Didn't take very long to start happening.  Data is still flowing into ES though.. so far.

**Comment by BrianMBeaulieu** [comment]
Eventually got an exception.. not sure if it's related
```
[2017-11-29T13:58:22,271][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input.
[2017-11-29T13:58:22,382][WARN ][io.netty.channel.AbstractChannelHandlerContext] An exception 'java.lang.NullPointerException' [enable DEBUG level for full stacktrace] was thrown by a user handler's exceptionCaught() method while handling the following exception:
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method) ~[?:1.8.0_141]
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39) ~[?:1.8.0_141]
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223) ~[?:1.8.0_141]
	at sun.nio.ch.IOUtil.read(IOUtil.java:192) ~[?:1.8.0_141]
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380) ~[?:1.8.0_141]
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1100) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:349) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:112) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:571) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:512) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:426) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:398) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:877) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_141]
[2017-11-29T13:58:22,383][WARN ][io.netty.channel.AbstractChannelHandlerContext] An exception 'java.lang.NullPointerException' [enable DEBUG level for full stacktrace] was thrown by a user handler's exceptionCaught() method while handling the following exception:
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method) ~[?:1.8.0_141]
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39) ~[?:1.8.0_141]
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223) ~[?:1.8.0_141]
	at sun.nio.ch.IOUtil.read(IOUtil.java:192) ~[?:1.8.0_141]
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380) ~[?:1.8.0_141]
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1100) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:349) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:112) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:571) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:512) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:426) ~[netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:398) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:877) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_141]

```

**Comment by ph** [comment]
@robbavey I was able to reproduce this error when looking at another problem, when using mutual auth certificate with a bad certificate you hit that **NullException**.

**Comment by ph** [comment]
I have added a null check for the remote host, I was only able to reproduce it sporadically by creating/shutting down repeatedly.

```
[2017-11-30T14:21:11,970][DEBUG][org.logstash.netty.SslSimpleBuilder] Ciphers:  [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256]
[2017-11-30T14:21:11,974][DEBUG][org.logstash.netty.SslSimpleBuilder] TLS: [TLSv1, TLSv1.1, TLSv1.2]
[2017-11-30T14:21:11,977][DEBUG][io.netty.handler.ssl.OpenSslEngine] SSL_read failed: OpenSSL error: error:10000412:SSL routines:OPENSSL_internal:SSLV3_ALERT_BAD_CERTIFICATE
[2017-11-30T14:21:11,978][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5044, remote: xxx.xxx.xxx.xxx.XXXX] Exception: javax.net.ssl.SSLHandshakeException: error:10000412:SSL routines:OPENSSL_internal:SSLV3_ALERT_BAD_CERTIFICATE
[2017-11-30T14:21:13,983][DEBUG][org.logstash.netty.SslSimpleBuilder] Available ciphers:[ECDHE-ECDSA-AES128-GCM-SHA256, ECDHE-RSA-AES128-GCM-SHA256, ECDHE-PSK-AES128-GCM-SHA256, DHE-RSA-AES128-GCM-SHA256, ECDHE-ECDSA-AES256-GCM-SHA384, ECDHE-RSA-AES256-GCM-SHA384, ECDHE-PSK-AES256-GCM-SHA384, DHE-RSA-AES256-GCM-SHA384, ECDHE-ECDSA-CHACHA20-POLY1305, ECDHE-RSA-CHACHA20-POLY1305, ECDHE-PSK-CHACHA20-POLY1305, ECDHE-ECDSA-AES128-SHA, ECDHE-ECDSA-AES128-SHA256, ECDHE-RSA-AES128-SHA, ECDHE-RSA-AES128-SHA256, ECDHE-PSK-AES128-CBC-SHA, DHE-RSA-AES128-SHA, DHE-RSA-AES128-SHA256, ECDHE-ECDSA-AES256-SHA, ECDHE-ECDSA-AES256-SHA384, ECDHE-RSA-AES256-SHA, ECDHE-RSA-AES256-SHA384, ECDHE-PSK-AES256-CBC-SHA, DHE-RSA-AES256-SHA, DHE-RSA-AES256-SHA256, ECDHE-ECDSA-RC4-SHA, ECDHE-RSA-RC4-SHA, AES128-GCM-SHA256, AES256-GCM-SHA384, AES128-SHA, AES128-SHA256, PSK-AES128-CBC-SHA, AES256-SHA, AES256-SHA256, PSK-AES256-CBC-SHA, DES-CBC3-SHA, RC4-SHA, PSK-RC4-SHA, RC4-MD5]
[2017-11-30T14:21:13,983][DEBUG][org.logstash.netty.SslSimpleBuilder] Ciphers:  [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256]
[2017-11-30T14:21:13,987][DEBUG][org.logstash.netty.SslSimpleBuilder] TLS: [TLSv1, TLSv1.1, TLSv1.2]
[2017-11-30T14:21:13,991][DEBUG][io.netty.handler.ssl.OpenSslEngine] SSL_read failed: OpenSSL error: error:10000412:SSL routines:OPENSSL_internal:SSLV3_ALERT_BAD_CERTIFICATE
[2017-11-30T14:21:13,991][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5044, remote: 127.0.0.1:52521] Exception: javax.net.ssl.SSLHandshakeException: error:10000412:SSL routines:OPENSSL_internal:SSLV3_ALERT_BAD_CERTIFICATE
[2017-11-30T14:21:15,965][DEBUG][logstash.outputs.file    ] Starting stale files cleanup cycle {:files=>{}}
[2017-11-30T14:21:15,965][DEBUG][logstash.outputs.file    ] 0 stale files found {:inactive_files=>{}}
[2017-11-30T14:21:17,995][DEBUG][org.logstash.netty.SslSimpleBuilder] Available ciphers:[ECDHE-ECDSA-AES128-GCM-SHA256, ECDHE-RSA-AES128-GCM-SHA256, ECDHE-PSK-AES128-GCM-SHA256, DHE-RSA-AES128-GCM-SHA256, ECDHE-ECDSA-AES256-GCM-SHA384, ECDHE-RSA-AES256-GCM-SHA384, ECDHE-PSK-AES256-GCM-SHA384, DHE-RSA-AES256-GCM-SHA384, ECDHE-ECDSA-CHACHA20-POLY1305, ECDHE-RSA-CHACHA20-POLY1305, ECDHE-PSK-CHACHA20-POLY1305, ECDHE-ECDSA-AES128-SHA, ECDHE-ECDSA-AES128-SHA256, ECDHE-RSA-AES128-SHA, ECDHE-RSA-AES128-SHA256, ECDHE-PSK-AES128-CBC-SHA, DHE-RSA-AES128-SHA, DHE-RSA-AES128-SHA256, ECDHE-ECDSA-AES256-SHA, ECDHE-ECDSA-AES256-SHA384, ECDHE-RSA-AES256-SHA, ECDHE-RSA-AES256-SHA384, ECDHE-PSK-AES256-CBC-SHA, DHE-RSA-AES256-SHA, DHE-RSA-AES256-SHA256, ECDHE-ECDSA-RC4-SHA, ECDHE-RSA-RC4-SHA, AES128-GCM-SHA256, AES256-GCM-SHA384, AES128-SHA, AES128-SHA256, PSK-AES128-CBC-SHA, AES256-SHA, AES256-SHA256

**Comment by ph** [comment]
From the Netty doc is possible to receive to receive null with both local address and remote address. Not sure how it can happen in our context.

```java
    /**
     * Returns the local address where this channel is bound to.  The returned
     * {@link SocketAddress} is supposed to be down-cast into more concrete
     * type such as {@link InetSocketAddress} to retrieve the detailed
     * information.
     *
     * @return the local address of this channel.
     *         {@code null} if this channel is not bound.
     */
    SocketAddress localAddress();

    /**
     * Returns the remote address where this channel is connected to.  The
     * returned {@link SocketAddress} is supposed to be down-cast into more
     * concrete type such as {@link InetSocketAddress} to retrieve the detailed
     * information.
     *
     * @return the remote address of this channel.
     *         {@code null} if this channel is not connected.
     *         If this channel is not connected but it can receive messages
     *         from arbitrary remote addresses (e.g. {@link DatagramChannel},
     *         use {@link DatagramPacket#recipient()} to determine
     *         the origination of the received message as this method will
     *         return {@code null}.
     */
    SocketAddress remoteAddress();
```

**Comment by robbavey** [comment]
@ph I [found a couple of places](https://github.com/logstash-plugins/logstash-input-beats/pull/270#issuecomment-346485562) in the Netty source where null can be returned for `remoteAddress` - the Windows case was interesting - as [one of the cases](https://github.com/logstash-plugins/logstash-input-beats/issues/269#issuecomment-347965715) mentions Windows boxes.

**Comment by BrianMBeaulieu** [comment]
I've been running in DEBUG and of course it hasn't happened since yesterday..  
9G of log file..

`
[elastic@es-broker logstash]$ grep 'Could not retrieve remote IP address for beats input' nohup.out 
[....]
[2017-11-29T14:42:38,724][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input.
[2017-11-29T14:42:38,783][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input.

[elastic@es-broker logstash]$ du -sm nohup.out 
9190	nohup.out
`

**Comment by sc-chad** [comment]
@robbavey 

We are exclusively Amazon Linux all within AWS.
All sources are from filebeat the agent is 5.6.4.

```
Installed Packages
Name        : filebeat
Arch        : x86_64
Version     : 5.6.4
Release     : 1
Size        : 29 M
Repo        : installed
From repo   : beats
Summary     : Filebeat sends log files to Logstash or directly to Elasticsearch.
URL         : https://www.elastic.co/products/beats/filebeat
License     : ASL 2.0
Description : Filebeat sends log files to Logstash or directly to Elasticsearch.
```

Logstash is behind an ALB all filebeat agents are pointing to the ALB via DNS.

Elasticsearch is configured in a 7 node cluster.  4 master nodes (no data) reside on the same logstash nodes, while 3 data nodes are isolated to their own instances.

Logstash communicates via localhost to the elasticsearch master installed locally on the same server.

Logstash is installed on 4 c5.xlarge instances.

**Comment by robbavey** [comment]
This should be fixed in v5.0.5 of the beats plugin.

**Comment by jensritter** [comment]
With the  logstash Version: 1:6.1.1-1 (which should be the curent one)  I am getting this also. 

After editing the file
`/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-input-beats-5.0.4-java/lib/logstash/inputs/beats/message_listener.rb`
 i coud reduce the logging, but still get "local: 0.0.0.0:5044, remote: undefined] Exception: null"

**Comment by robbavey** [comment]
@jensritter Do you still get this if you upgrade the beats input to 5.0.5?

**Comment by jensritter** [comment]
no - after upgrading to 5.0.5 it now works.--  sorry for the bug-spam.

**Comment by robbavey** [comment]
Closing this issue as it should be fixed by upgrading the beats input plugin to `5.0.5`

**Comment by agolomoodysaada** [comment]
Is there any docs on upgrading the input plugin?

**Comment by jensritter** [comment]
use the bin/logstash-plugin command to download and install the new version
from the internet.

https://www.elastic.co/guide/en/logstash/current/working-with-plugins.html

**Comment by mchesler** [comment]
I've upgraded the plugin to version `5.0.5`, and I'm still seeing the error mentioned by @jensritter:
`[2017-12-29T16:07:06,445][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5043, remote: undefined] Exception: null`

I'm also seeing this other error:
`[2017-12-29T16:08:27,163][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5043, remote: undefined] Exception: (NoMethodError) undefined method `accept' for nil:NilClass`

It's unclear to me if this is a related or different issue.

**Comment by robbavey** [comment]
@mchesler This is a separate issue - would you mind creating a new issue, and giving as much detail as you can in the issue? How regularly is this error message appearing?

Thanks

**Comment by sc-chad** [comment]
I wouldn't close this issue just yet @robbavey.

It looks like the original issue has now been replace with:
```
[2018-01-02T21:45:56,605][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5044, remote: undefined] Exception: null
```

I'm getting this on just about every message that comes through.  This was after upgrading to the latest plugin version.

**Comment by robbavey** [comment]
@sc-chad Are you also seeing the ```[2017-12-29T16:08:27,163][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5043, remote: undefined] Exception: (NoMethodError) undefined methodaccept' for nil:NilClass``` issue?

**Comment by sc-chad** [comment]
@robbavey Yes I am.  Although it isn't near as frequent as the others.

Also there is an oddity that some report the local one way while others report it another.

Example:
```
[2018-01-02T22:05:41,066][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5044, remote: undefined] Exception: null
[2018-01-02T22:05:43,156][INFO ][org.logstash.beats.BeatsHandler] [local: 10.0.5.126:5044, remote: 10.0.1.109:26508] Exception: null
[2018-01-02T22:05:43,171][INFO ][org.logstash.beats.BeatsHandler] [local: 10.0.5.126:5044, remote: 10.0.9.51:15691] Exception: (NoMethodError) undefined method `accept' for nil:NilClass
[2018-01-02T22:05:43,172][INFO ][org.logstash.beats.BeatsHandler] [local: 0.0.0.0:5044, remote: undefined] Exception: null
```

**Comment by sc-chad** [comment]
@robbavey Not sure if I'm doing this right honestly, but thought I'd take a shot.

So I changed: https://github.com/logstash-plugins/logstash-input-beats/blob/master/lib/logstash/inputs/beats/message_listener.rb#L96

To:
```
if (defined?(remote_address)).nil?
```

I now get:
```
[2018-01-02T22:30:50,416][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input. {:error=>#<NoMethodError: undefined method `getAddress' for nil:NilClass>}
[2018-01-02T22:30:50,822][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input. {:error=>#<NoMethodError: undefined method `getAddress' for nil:NilClass>}
[2018-01-02T22:30:51,660][WARN ][logstash.inputs.beats    ] Could not retrieve remote IP address for beats input. {:error=>#<NoMethodError: undefined method `getAddress' for nil:NilClass>}```

If that helps any.

**Comment by robbavey** [comment]
Thanks @sc-chad - unfortunately the change you added here always resolves to `false`, which is causing the line at https://github.com/logstash-plugins/logstash-input-beats/blob/master/lib/logstash/inputs/beats/message_listener.rb#L100 to always fire regardless of whether `remote_address` is `nil` - hence the ```undefined method `getAddress' for nil:NilClass>``` in your error message.

One thing I am interested in - when you see the `undefined method `accept' for nil:NilClass` error, are there other errors in the logs - not necessarily from the beats input?

**Comment by sc-chad** [comment]
@robbavey Had a feeling that's what it would do.

So oddly, after leaving everything up for a day the errors have stopped.  Haven't changed anything (using default code).

Makes me wonder if these errors only occur under very high load or when filebeat has many messages queued on the clients.

Scenario that would describe yesterday is logstash was down for a few hours so the filebeat clients were backed up on messages in queue.  

Once we started logstash it gets slammed from all the pending messages which results in those errors.  Now that it's caught up and the load is back to normal, we aren't getting these messages anymore.

**Comment by jensritter** [comment]
I can confirm that this error occurs during heavy load.

 I am currently deploying winlogbeat to old workstations - and get this error, during the initial bulk-import per client.

**Comment by robbavey** [comment]
The cause for this appears to be #282

**Comment by sc-chad** [comment]
@robbavey Awesome nice job!

**Comment by mchesler** [comment]
Thanks for getting that change in so quickly @robbavey!! Any idea how long before a new version of the plugin is available?

**Comment by robbavey** [comment]
@mchesler New versions of the plugin have been released - `5.0.6` for Logstash 6.x and the fix has also been backported to `3.1.26` for Logstash 5.x

**Comment by mchesler** [comment]
Awesome! Thanks @robbavey!

**Comment by robbavey** [comment]
Closing this issue again - #283 should have fixed the spamming of logs with `Exception: null`

**Comment by mchesler** [comment]
Confirmed.  I have not seen the `Exception: null` error since updating to version `5.0.6`
